### PR TITLE
ThemeSheet: remove "Try & Customize" button

### PR DESCRIPTION
**Pull request by sirbrillig**
This is for use after the Theme Sheets are launched.

The "Try & Customize" button in the theme sheet preview is problematic for a few reasons, but primarily IMO it can cause a lot of confusion with Headstart.

When Headstart is fully launched (soon), switching themes on a site with no custom content (basically a new site) will cause all the content on a site (posts, pages, menus, widgets) to be replaced with placeholder content appropriate to the theme.

If a user presses "Try & Customize" prior to switching themes, they will see and customize their old content, and then when they activate the switch, all those customizations will be replaced by the new placeholder content, making for a very poor experience.

With this PR, anyone viewing a site preview through Theme Sheets or the theme showcase will not have the option of "Try & Customize", but will go directly to "Activate".

That is, when clicking this button:

screen_shot_2016-06-14_at_1_15_04_pm

Before:

screen shot 2016-06-14 at 1 14 51 pm

After:

screen shot 2016-06-24 at 11 17 39 am

## Testing Instructions

(trying to cover all the things this PR touches)

1. Visit the theme sheet page (eg: http://calypso.localhost:3000/theme/twentysixteen) and verify that the button looks ok and works correctly ("Pick this design" and "Free" or the price).
2. Try activating the theme and make sure it brings up the site picker and the theme is activated.
3. Repeat steps 1-2 for the theme sheet page for a specific site (eg: http://calypso.localhost:3000/theme/twentysixteen/mysite.wordpress.com)
4. Same test for the theme sheet page for a specific site, but use the theme sheet for a site that already has that theme. The button should read "Customize" instead (with no price).
5. Repeat steps 1-4 after clicking on the "Open live demo" link in the theme sheet to bring up the theme preview.
6. Visit the theme showcase page (eg: http://calypso.localhost:3000/design/) and click on a theme with a price, verifying that the button in the preview says "Pick this design" and has the price.
7. Repeat step 5 with a free theme and verify that it reads "Free".
8. Visit the theme showcase page for a specific site (eg: http://calypso.localhost:3000/design/mysite.wordpress.com) and repeat step 7.
9. Activate a theme after step 8 and verify that it is activated.
10. Visit the theme showcase while logged-out (eg: http://calypso.localhost:3000/design/) and repeat step 9, but verify that clicking the link brings you to sign up and the resulting site has the theme you chose.

**In summary**, this button changes or updates the UI for:
- button variations (free/paid themes, the currently active theme)
- theme sheet (single site, multi-site, logged-out, and associated theme previews)
- theme showcase previews (single site, multi-site, and logged-out)

Test live: https://calypso.live/?branch=remove/theme-sheet-customize-button

### CODE REVIEW

**Comment by folletto** [comment]
Since it's functionally equivalent, any way this could match the other button visually (white, green price)?

screen shot 2016-06-14 at 19 20 47

**Comment by sirbrillig** [comment]
@folletto yes, that's a good idea.

**Comment by sirbrillig** [comment]
This is a little tricky because the reason the button is blue is that it is marked as `primary`. If we change it to non-primary, it will also affect the logged-out preview:

screen shot 2016-06-14 at 4 58 15 pm

That might be fine (and it would certainly be consistent).

Also worth considering is that if the button is not primary, it can get a little lost in all the white on the preview page:

screen shot 2016-06-14 at 5 00 00 pm

Adding the bright green price might help with that, but I don't know if it is a better experience than the blue button. 

Thoughts?

**Comment by folletto** [comment]
I agree it would be less visible, but I think it's fine considering the user won't simply "drop" into this screen, but would get there from the previous page, which is either the full theme sheet or the showcase. :)

Also, the green bit (Free, or price) adds a tiny bit of color that helps standing out a bit more. :)

**Comment by sirbrillig** [comment]
Added the price.

screen shot 2016-06-16 at 3 26 21 pm

**Comment by folletto** [comment]
Tested on:

* Free theme
* Premium theme
* Active theme

Works well, with one exception: on the active theme it still shows "Pick this design" instead of "Customize".

---

I was wondering... should we use that also from the main showcase list? (currently shows "Try and Customize" stil). I think we can leave "Try and Customize" only inside the ••• menu. Thoughts?

**Comment by sirbrillig** [comment]
> Works well, with one exception: on the active theme it still shows "Pick this design" instead of "Customize".

Whoops! Must be a bug. I'll figure that out.

> I was wondering... should we use that also from the main showcase list? (currently shows "Try and Customize" stil). I think we can leave "Try and Customize" only inside the ••• menu

IMO we should remove every instance of "Try and Customize", or at least remove every instance of it if the user's site would be Headstarted. Thanks to some API changes we can know that inside Calypso, but unfortunately that doesn't help if we're looking at themes for multiple sites. My uncreative solution is to just remove it everywhere, 😛 but I'd like to hear other ideas.

**Comment by folletto** [comment]
Ok!

Let's remove it then from the overlay in the list. :)

I'd still lean on keeping it inside ••• since it might be an advanced option useful to some. Would that be a problem? Can we get a site to go live on that switch with the changes done by the users and without Headstart?

**Comment by sirbrillig** [comment]
> Can we get a site to go live on that switch with the changes done by the users and without Headstart?

Oh, that's an idea I hadn't considered! There may indeed be a way. All we would need to do is remove the blog sticker that shows the site hasn't been modified. Let me discuss this with the team.

**Comment by sirbrillig** [comment]
@mattwiebe had an idea for disabling Headstart if we activate at theme via the Customizer. I think that would take care of the "advanced" option of "try and customize" from the ellipses menu. Assuming that works, I'll make the changes suggested above.

**Comment by folletto** [comment]
Sounds good. 👍

**Comment by folletto** [comment]
I was chatting with Steve and we considered that it's probably ideal if the "Pick" button:

screen shot 2016-06-21 at 15 58 39

Was a separate component, so it's consistently rendered everywhere with the price and all. :)

**Comment by mattwiebe** [comment]
FYI theme switches activated via the Customizer will now not be Headstarted.

**Comment by folletto** [comment]
For reference, this can land after #6221 is closed.

**Comment by sirbrillig** [comment]
For my reference, current to do (probably):

- [ ] Make sure preview display for active theme displays "Customize"
- [ ] Move "Pick" button to its own component
- [ ] Replace "Try & Customize" in all instances
- [ ] Add "Try & Customize" to ... menu

**Comment by folletto** [comment]
> Add "Try & Customize" to ... menu

This should be already covered in #6249, ready for review. ;)

**Comment by sirbrillig** [comment]
Creating `ThemePickButton` ended up being a bigger project than I expected. After several revisions I decided (but am still open to discussion) to make it a Container component, giving it access to the state tree so it can get its own props and call its own actions rather than trying to pass in everything that it needs from above. The result is that the only required prop is `theme`, and the button takes care of everything else. 

One additional advantage to this is that we get to remove a ton of code from the components that contain the button.

Initial tests seem positive, but we should test this heavily since the button is in many different places and can have several different effects (customize, activate, purchase, sign up).

**Comment by folletto** [comment]
> Initial tests seem positive, but we should test this heavily since the button is in many different places and can have several different effects (customize, activate, purchase, sign up).

I still see this tagged "In Progress" but from the message seems more a "Needs Review"... should we switch the tag and open for testing and review?

**Comment by folletto** [comment]
I tested briefly on chrome. From the list it works, but trying to open the theme sheet throws:

```
TypeError: Cannot read property 'showPreview' of null
   at [object Object].renderSheet (/home/ubuntu/data/gh-wp-calypso/branches/remove/theme-sheet-customize-button/build/webpack:/client/my-sites/theme/main.jsx:279:7)
   at [object Object].render (/home/ubuntu/data/gh-wp-calypso/branches/remove/theme-sheet-customize-button/build/webpack:/client/my-sites/theme/main.jsx:307:15)
   at [object Object].ReactCompositeComponentMixin._renderValidatedComponentWithoutOwnerOrContext (/home/ubuntu/data/gh-wp-calypso/branches/remove/theme-sheet-customize-button/build/webpack:/~/react/lib/ReactCompositeComponent.js:815:1)
   at [object Object].ReactCompositeComponentMixin._renderValidatedComponent (/home/ubuntu/data/gh-wp-calypso/branches/remove/theme-sheet-customize-button/build/webpack:/~/react/lib/ReactCompositeComponent.js:841:1)
   at [object Object].ReactCompositeComponentMixin.performInitialMount (/home/ubuntu/data/gh-wp-calypso/branches/remove/theme-sheet-customize-button/build/webpack:/~/react/lib/ReactCompositeComponent.js:351:1)
   at [object Object].ReactCompositeComponentMixin.mountComponent (/home/ubuntu/data/gh-wp-calypso/branches/remove/theme-sheet-customize-button/build/webpack:/~/react/lib/ReactCompositeComponent.js:244:1)
   at Object.ReactReconciler.mountComponent (/home/ubuntu/data/gh-wp-calypso/branches/remove/theme-sheet-customize-button/build/webpack:/~/react/lib/ReactReconciler.js:46:1)
   at [object Object].ReactCompositeComponentMixin.performInitialMount (/home/ubuntu/data/gh-wp-calypso/branches/remove/theme-sheet-customize-button/build/webpack:/~/react/lib/ReactCompositeComponent.js:357:1)
   at [object Object].ReactCompositeComponentMixin.mountComponent (/home/ubuntu/data/gh-wp-calypso/branches/remove/theme-sheet-customize-button/build/webpack:/~/react/lib/ReactCompositeComponent.js:244:1)
   at Object.ReactReconciler.mountComponent (/home/ubuntu/data/gh-wp-calypso/branches/remove/theme-sheet-customize-button/build/webpack:/~/react/lib/ReactReconciler.js:46:1)
```

**Comment by sirbrillig** [comment]
Whoops, I accidentally broke the theme sheet when I was cleaning up last night. Try it now.

**Comment by sirbrillig** [comment]
Added (rather long) testing instructions.

**Comment by sirbrillig** [comment]
Much of this code (and testing) will become simpler after #6342

**Comment by folletto** [comment]
Ok! #6342 (#6221) and #6249 got merged, so we have no blockers for this anymore.

File: client/my-sites/theme/main.jsx [hunk 1]
```diff
@@ -71,48 +67,17 @@ const ThemeSheet = React.createClass( {
...
-	},
-
-	selectSiteAndDispatch( action ) {
-		if ( this.props.selectedSite ) {
-			this.props[ action ]( this.props, this.props.selectedSite, 'showcase-sheet' );
-		} else {
-			this.setState( { selectedAction: action } );
-		}
+	getTheme() {
+		return pick( this.props, [ 'id', 'name', 'author', 'screenshot', 'price', 'description', 'descriptionLong', 'supportDocumentation', 'download', 'taxonomies', 'stylesheet', 'active', 'purchased' ] );
```
Comment (seear) [inline client/my-sites/theme/main.jsx]: Is there a benefit to this `pick`? Seems it has a drawback, in that it may need to be altered from time to time.

File: client/my-sites/theme/theme-pick-button.jsx [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+	renderSiteSelector() {
+		if ( ! this.state.selectedAction ) {
+			return;
+		}
+		const selectedAction = this.state.selectedAction;
+		const onCompleteSiteSelection = ( selectedTheme, site ) => {
+			this.props[ selectedAction ]( selectedTheme, site, 'showcase-sheet' );
+			this.props.onClick( this.props.theme );
+		};
+		// TODO: this used to be `section`, but I don't know what that is or if it
```
Comment (seear) [inline client/my-sites/theme/theme-pick-button.jsx]: `section` is the current theme sheet page, one of _Overview/Setup/Support_. It won't be needed in this component.

**Comment by ockham** [comment]
Giving this another read, there's one overlap of concerns with #6349 and friends. Over there, my idea was to stick to the concept of an `options` object (with label and actions, much like `themes/single-site` et al currently have in their `getButtonOptions()` method), pick a default option from that object, and pass it to the button. (This would keep the button a little dumber, and de-deupe some logic.)

I think the most relevant case to look at is the SiteSelector stuff which I think can be more cleanly handled by the HoC in #6349 than we're now doing. We might need to give some consideration how we could use that HoC with the approach in this PR. (Wrapping the button would duplicate the wrapping for  `/themes/multi-site`; wrapping `theme/main` would probably break encapsulation of logic inside the button :-/ )

**Comment by seear** [comment]
I was initially surprised by just how much logic has moved from the main theme sheet component into the new button component. Thinking it through, is seems like this is the right way to go. It is right that `main.jsx` should become less complex over time, and now we get to share the same logic in multiple places.

I'll come back with a more thorough review and testing after the pending rebases.

**Comment by ockham** [comment]
I'm now really leaning towards thinking we should take the abstraction one level higher, which is by passing an option object to the theme sheet which holds button label and action, and wrapping the sheet in the HoC from #6349 so it works nicely with the site selector.

As for the option object, I'm thinking we can re-use the various options from `single-site`, `multi-site` and `logged-out` (to contain the logic) and pick the respective default action for each case for the theme sheet pick button, again probably at `connect` stage.

File: client/my-sites/theme/main.jsx (hunk 1 above)
Comment (sirbrillig) [inline client/my-sites/theme/main.jsx]: I don't like it either. Perhaps there's a better way. The intention was to re-create the `theme` object that was deconstructed and injected into this component; there are so many parts of that object that are needed by the ThemePickButton (and its children like the site picker) that it seemed overkill to try to specify each one, and imprecise to just pass `this.props` along since that doesn't describe what the component does. 

In my ideal world this component would be passed `theme` rather than deconstructed pieces, and we can just pass that along to the button, but I didn't want to make this PR any bigger than necessary.

Comment (seear) [inline client/my-sites/theme/main.jsx]: Yeah, we've struggled with this in a number of places. It would be great to have a common `theme` object. One obstacle to that is that different endpoints return different fields. 

> imprecise to just pass this.props along since that doesn't describe what the component needs.

I don't see a huge problem with passing `this.props`. The proptypes in the component can describe the needs. I admit it can be confusing at the site where the component is used, though, so your way is fine.

**Comment by sirbrillig** [comment]
To quote from a discussion with @ockham on Slack:

>  the ThemePickButton would still be used in the ThemeSheet and the previews for logged-out, single-site, multi-site, and ThemeSheet, but the button would not need all the action handler code - the necessary action would be passed in as a prop. The decision about which action to show would be made by the parent or a component higher up the tree.

So I'll wait until #6349 is complete before refactoring this.

File: client/my-sites/theme/main.jsx (hunk 1 above)
Comment (sirbrillig) [inline client/my-sites/theme/main.jsx]: > It would be great to have a common theme object. One obstacle to that is that different endpoints return different fields.

That's good info. Looking at it again, my concern is the dependencies (eg: `customize()` from `state/themes/actions`, `isPremium()`, `selectedTheme` prop from `ThemesSiteSelectorModal`) which accept a theme object. IMO we should probably either standardize the theme object for these things or have them accept simpler props (eg: just theme slug).

I may be perpetuating the issue here rather than simplifying it (as I had intended). When I refactor this I'll look at each of those dependencies and see if I can construct simple objects for each one with just the data they need; then I'll be able to determine what data the button needs and specify that explicitly in the PropTypes.

Comment (ockham) [inline client/my-sites/theme/main.jsx]: > Looking at it again, my concern is the dependencies (eg: `customize()` from `state/themes/actions`, `isPremium()`, `selectedTheme` prop from `ThemesSiteSelectorModal`) which accept a theme object. IMO we should probably either standardize the theme object for these things or have them accept simpler props (eg: just theme slug).

While we won't be able to standardize all function signatures (e.g. have all functions from that group accept exactly one theme and one site argument), I've started a PR to turn our theme helpers into selectors (which I believe they ultimately are), and which will clean up things a bit as it will enable us to bind to `site` early, i.e. at `connect` stage.

Tangentially, I'd like then to change all theme selectors to accept theme IDs (instead of entire objects), as that will make them more versatile (e.g for edge cases like [this](https://github.com/Automattic/wp-calypso/blob/8a20ad81021ebb87bef3b3b4aac341b209c6caca/client/me/purchases/product-link/index.jsx#L41)). I think that style is starting to become accepted practice (cf site related selectors) -- if really needed, a selector can always get the corresponding object itself.

**Comment by sirbrillig** [comment]
I think this is resolved by #6349 (in a different way).
